### PR TITLE
Add workflow safe Mutex and Semaphore

### PR DESCRIPTION
**Pull request by Quinn-With-Two-Ns**
Add workflow safe Mutex and Semaphore. Based on Go standard [Mutex](https://pkg.go.dev/sync) and [Semaphore](https://pkg.go.dev/golang.org/x/sync/semaphore). Notable difference for the `Mutex` is our mutex supports cancellation of a lock request. I did this to match `Semaphore` and most other Go SDK primitives.

### CODE REVIEW

File: internal/internal_workflow.go [hunk 1]
```diff
@@ -1742,3 +1754,56 @@ func (us updateSchedulerImpl) Spawn(ctx Context, name string, highPriority bool,
...
+func (s *semaphoreImpl) TryAcquire(ctx Context, n int64) bool {
+	if s.count < n {
+		return false
+	}
+	s.count -= n
+	return true
+}
+
+func (s *semaphoreImpl) Release(n int64) {
+	s.count += n
```
Comment (cretz) [inline internal/internal_workflow.go:1808]: Arguably could have a panic here if count exceeds the initial number given at construction

File: internal/internal_workflow.go [hunk 2]
```diff
@@ -1742,3 +1754,56 @@ func (us updateSchedulerImpl) Spawn(ctx Context, name string, highPriority bool,
...
+}
+
+func (m *mutexImpl) Unlock() {
+	if !m.locked {
+		panic("Mutex.Unlock() was called on an unlocked mutex")
+	}
+	m.locked = false
+}
+
+func (m *mutexImpl) IsLocked() bool {
```
Comment (cretz) [inline internal/internal_workflow.go:1784]: Would be ok if you didn't expose this, but up to you

File: internal/internal_workflow.go [hunk 3]
```diff
@@ -1742,3 +1754,56 @@ func (us updateSchedulerImpl) Spawn(ctx Context, name string, highPriority bool,
...
+		return !m.locked
+	})
+	if err != nil {
+		return err
+	}
+	m.locked = true
+	return nil
+}
+
+func (m *mutexImpl) TryLock(ctx Context) bool {
```
Comment (cretz) [inline internal/internal_workflow.go:1769]: Arguably don't need a context here, but I don't mind it, so long as it's understood that calling `Lock` with a cancelled context is different than calling `TryLock` with a cancelled context.

**Review by cretz — APPROVED** [review]
LGTM. You are lucky that only one awaiter is ever awoken at a time, heh.

File: internal/internal_workflow.go (hunk 1 above)
Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow.go:1808]: Let me double check what Go's semaphore does, I don't think it cares

File: internal/internal_workflow.go (hunk 2 above)
Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow.go:1784]: I exposed it because I wanted a condition a user could check in an `Await` or update validator

File: internal/internal_workflow.go (hunk 3 above)
Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow.go:1769]: Ah I added a context here so I could add a check this wasn't being called in an read only context.

File: internal/internal_workflow.go (hunk 1 above)
Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow.go:1808]: AH yes it should panic

Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow.go:1809]: Added a panic here

File: internal/internal_workflow.go [hunk 4]
```diff
@@ -1742,3 +1755,61 @@ func (us updateSchedulerImpl) Spawn(ctx Context, name string, highPriority bool,
...
+	if m.locked {
+		return false
+	}
+	m.locked = true
+	return true
+}
+
+func (m *mutexImpl) Unlock() {
+	if !m.locked {
+		panic("Mutex.Unlock() was called on an unlocked mutex")
```
[the remark this answers is not in the document]
Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow.go:1781]: Yes, this is programming error, same as accessing a null value or access an array out of bounds which all panic and is consistent with how Go's native mutext behaves

[the remark this answers is not in the document]
Comment (Sushisource) [reply internal/internal_workflow.go:1781]: Sure, of course that's true, I'm just thinking maybe we want to afford some different control for our alternatives given that our environment is different from a normal programming environment

[the remark this answers is not in the document]
Comment (Quinn-With-Two-Ns) [reply internal/internal_workflow.go:1781]: I think in our environment it is even more important this is a panic vs an error that users could unintentionally handle  incorrectly.

File: internal/internal_workflow.go [hunk 5]
```diff
@@ -1742,3 +1755,61 @@ func (us updateSchedulerImpl) Spawn(ctx Context, name string, highPriority bool,
...
+	if success {
+		s.cur += n
+	}
+	return success
+}
+
+func (s *semaphoreImpl) Release(n int64) {
+	s.cur -= n
+	if s.cur < 0 {
+		panic("Mutex.Release() released more than held")
```
Comment (dandavison) [inline internal/internal_workflow.go:1813]:
```suggestion
		panic("Semaphore.Release() released more than held")
```

File: internal/internal_workflow.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sushisource) [reply internal/internal_workflow.go:1781]: Sure, that's probably true. Works for me then. I think it's worth at least considering the options here but I agree it sounds like there's not a great case for making them return errors instead.

**Review by Sushisource — APPROVED** [review]

File: test/integration_test.go [hunk 6]
```diff
@@ -1562,6 +1562,122 @@ func (ts *IntegrationTestSuite) TestUpdateWorkflowCancelled() {
...
+		WorkflowID:   run.GetID(),
+		RunID:        run.GetRunID(),
+		UpdateName:   "update",
+		WaitForStage: client.WorkflowUpdateStageAccepted,
+		Args:         []interface{}{false},
+	})
+	ts.NoError(err)
+	// Cancel the workflow, this should cancel any update blocking on the mutex
+	ts.NoError(ts.client.CancelWorkflow(ctx, run.GetID(), run.GetRunID()))
+	ts.Error(updateToCancel.Get(ctx, nil))
```
Comment (dandavison) [inline test/integration_test.go:1622]: Is this assertion checking that the blocking wait operation on a mutex is cancellable, or just checking that canceling a workflow cancels pending updates?

(I'm finding the tests a bit hard to follow; I think they would be easier to follow if there were a few smaller tests each testing something narrower.)

Comment (Quinn-With-Two-Ns) [reply test/integration_test.go:1622]: The former, the later is not a feature. I would leave the test as is since is is trying to test multiple updates at once competing for the  lock.
